pkg/security/probe/load_controller.go

```diff
@@ -45,31 +68,92 @@ func NewLoadController(probe *Probe, statsdClient *statsd.Client) (*LoadControll
 		return nil, err
 	}
 
+	forkLru, err := simplelru.NewLRU(probe.config.PIDCacheSize, nil)
+	if err != nil {
+		return nil, err
+	}
+
 	lc := &LoadController{
-		probe:                probe,
-		counters:             lru,
-		statsdClient:         statsdClient,
+		probe:        probe,
+		statsdClient: statsdClient,
+		hashFunc:     hashFunc,
+
+		eventsCounters: lru,
+		forkCounters:   forkLru,
+
 		EventsCountThreshold: probe.config.LoadControllerEventsCountThreshold,
+		ForkBombThreshold:    probe.config.LoadControllerForkBombThreshold,
 		DiscarderTimeout:     probe.config.LoadControllerDiscarderTimeout,
 		ControllerPeriod:     probe.config.LoadControllerControlPeriod,
 	}
 	return lc, nil
 }
 
-// Count increments the event counter of the provided event type and pid
-func (lc *LoadController) Count(eventType EventType, pid uint32) {
+// Count processes the provided events and ensures the load of the provided event type is within the configured limits
+func (lc *LoadController) Count(event *Event) {
+	switch event.GetEventType() {
+	case ExitEventType, ExecEventType, InvalidateDentryEventType:
+		return
+	case ForkEventType:
+		lc.CountFork(event)
+	default:
+		lc.GenericCount(event)
+	}
+}
+
+// CountFork increments the fork counter of the provided cookie
+func (lc *LoadController) CountFork(event *Event) {
 	lc.Lock()
 	defer lc.Unlock()
+	var forkBomb bool
+	var newCount uint64
 
-	entry, ok := lc.counters.Get(eventCounterLRUKey{Pid: pid, Event: eventType})
+	// About the key of the LRU below.
+	// - We could have used the process cookie. However cookies are not set in kernel space during the snapshot, which
+	//   means the forks of the existing processes on startup will all be counted at key "0". In other words this does
+	//   not work. (setting a cookie on snapshot doesn't work either because of possible races)
+	// - We could have used (mount_id, inode). With all the problems we've had with caches indexed by inodes, we decided
+	//   to stay on the safe side an avoid having to invalidate keys based on unlink / rename / ...
+	lruKey := newForkLRUKey(lc.hashFunc([]byte(event.Process.ResolveContainerPath(event) + "/" + event.Process.ResolveInode(event))))
+	entry, ok := lc.forkCounters.Get(lruKey)
+	if ok {
+		counter := entry.(*uint64)
+		if atomic.AddUint64(counter, 1) >= uint64(lc.ForkBombThreshold) {
+			forkBomb = true
+
+			// reset counter to avoid spamming fork bomb alerts until the best_effort move takes effect
+			atomic.SwapUint64(counter, 0)
+		}
+	} else {
+		newCount = uint64(1)
+		lc.forkCounters.Add(lruKey, &newCount)
+	}
+
+	if forkBomb {
+		if err := lc.statsdClient.Count(MetricForkBomb, 1, []string{}, 1); err != nil {
+			log.Warn(errors.Wrapf(err, "failed to send %s metric", MetricForkBomb))
+		}
+		lc.probe.DispatchCustomEvent(NewForkBombEvent(event, time.Now()))
```

Comment:
I will call `time.Now()` in the `marshall` function other we are going to generate of lot call for it that are just going to be rate limited

Reply:
What about `event.ResolveEventTimestamp()` ?

Reply:
(which we can also do in the `marshall` function)